### PR TITLE
Fixes #65377: get_datetime_* functions can return wrong values

### LINKED ISSUE #65377: get_datetime_* functions can return wrong values
### Godot version

00fa4e23e4b967291034ea7bb018dea638b37e8f

### System information

all OSes

### Issue description

This was first noticed by aaronfranke in another issue: https://github.com/godotengine/godot/issues/60327#issuecomment-1101453623

In https://github.com/godotengine/godot/blob/master/core/os/time.cpp this function is used to return datetime dict:

```c++
Dictionary Time::get_datetime_dict_from_system(bool p_utc) const {
	OS::Date date = OS::get_singleton()->get_date(p_utc);
	OS::Time time = OS::get_singleton()->get_time(p_utc);
	Dictionary datetime;
	datetime[YEAR_KEY] = date.year;
	datetime[MONTH_KEY] = (uint8_t)date.month;
	datetime[DAY_KEY] = date.day;
	datetime[WEEKDAY_KEY] = (uint8_t)date.weekday;
	datetime[DST_KEY] = date.dst;
	datetime[HOUR_KEY] = time.hour;
	datetime[MINUTE_KEY] = time.minute;
	datetime[SECOND_KEY] = time.second;
	return datetime;
}
```

Date and time parts are fetched using separate `get_date` and `get_time` calls. If the `get_date` call is made just before day ends (e.g. 2022-01-31 23:59:59.999999) and `get_time` call is made on the next day (e.g 2022-02-01 00:00:00.0000001) the resulted datetime dict will be 2022-01-31 00:00:00.0000001, which is wrong by 24 hours. The same problem is in `get_datetime_string_from_system` function.

This could be fixed by adding to `core/os/os.h` a new `get_datetime` function which would get both date and time using just one OS call.

### Steps to reproduce

I haven't actually seen this bug to happen and the bug would be quite hard to reproduce as the timing must perfect.

### Minimal reproduction project

_No response_

### PR DESCRIPTION
**Pull request by gotnospirit**
Add `OS::get_datetime` and use it when `OS::get_date` and `OS::get_time` were called together.

### CODE REVIEW

File: core/os/os.h [hunk 1]
```diff
@@ -216,13 +216,19 @@ class OS {
 		uint8_t second;
 	};
 
+	struct DateTime {
+		Date date;
+		Time time;
+	};
```
Comment (aaronfranke) [inline core/os/os.h:222]: I'm wondering if we should take this a step further, and do this:

```suggestion
	struct DateTime {
		int64_t year;
		Month month;
		uint8_t day;
		Weekday weekday;
		uint8_t hour;
		uint8_t minute;
		uint8_t second;
		bool dst;
	};
```

If so, we wouldn't need `OS.get_date()` and `OS.get_time()`, or the `Date` and `Time` structs. I'm not entirely sure if there are other uses of them that I'm missing though.

**Review by aaronfranke — COMMENTED** [review]
Overall this PR is good, it fixes an edge case and has the potential to make our API cleaner.

File: core/os/os.h (hunk 1 above)
Comment (gotnospirit) [reply core/os/os.h:222]: With this PR, both structs are only used in few `Time::get_X_from_system` methods, so replacing `OS.get_date` and `OS.get_time` is pretty straightforward.
I will update to the PR

File: core/os/time.cpp [hunk 2]
```diff
@@ -324,63 +324,61 @@ String Time::get_offset_string_from_offset_minutes(int64_t p_offset_minutes) con
...
-	date_dictionary[YEAR_KEY] = date.year;
-	date_dictionary[MONTH_KEY] = (uint8_t)date.month;
-	date_dictionary[DAY_KEY] = date.day;
-	date_dictionary[WEEKDAY_KEY] = (uint8_t)date.weekday;
-	date_dictionary[DST_KEY] = date.dst;
+	date_dictionary[YEAR_KEY] = dt.year;
+	date_dictionary[MONTH_KEY] = (uint8_t)dt.month;
+	date_dictionary[DAY_KEY] = dt.day;
+	date_dictionary[WEEKDAY_KEY] = (uint8_t)dt.weekday;
+	date_dictionary[DST_KEY] = dt.dst;
```
Comment (aaronfranke) [inline core/os/time.cpp:347]: Now that my [previous concern](https://github.com/godotengine/godot/issues/60327#issuecomment-1101453623) of discarding information is no longer relevant, I think we should remove DST from the date getter, and only include it in `get_datetime`.
```suggestion
```

Comment (gotnospirit) [reply core/os/time.cpp:347]: In v4, `get_datetime` isn't exposed to user anymore, all access to date informations was moved to `Time` class, so removing DST keys from `Time.get_X` will prevent user to any access to it

Comment (aaronfranke) [reply core/os/time.cpp:347]: Maybe I'm missing something here, but what I am saying is that we can include DST information in `Time::get_datetime_dict_from_system` and exclude it in `Time::get_date_dict_from_system`.

Comment (gotnospirit) [reply core/os/time.cpp:347]: My bad, you meant just removing the DST key in `get_date_dict_from_system`, right?

Comment (gotnospirit) [reply core/os/time.cpp]: Done! The [documentation](https://docs.godotengine.org/en/latest/classes/class_time.html#class-time-method-get-date-dict-from-system) should be updated

**Review by aaronfranke — APPROVED** [review]
Thanks for doing this work, this looks like a good improvement. 👍

I haven't tested locally but assuming it works this should fix #65377.

EDIT: Docs should be updated, and also this PR needs [a squash](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html).

File: core/os/time.cpp (hunk 2 above)
Comment (gotnospirit) [reply core/os/time.cpp]: Should I stash this 3 commits?

Comment (aaronfranke) [reply core/os/time.cpp]: Thanks, the documentation should be updated in this PR, in `doc/classes/Time.xml`.

**Review by aaronfranke — APPROVED** [review]

**Comment by gotnospirit** [comment]
I can submit a PR for 3.x but I'm wondering if the DST key should also be removed from `_OS::get_date`?

**Comment by gotnospirit** [comment]
I will update the PR again because the documentation of `Time.get_datetime_dict_from_system` doesn't mention the dst key

File: core/os/time.cpp [hunk 3]
```diff
@@ -324,63 +324,60 @@ String Time::get_offset_string_from_offset_minutes(int64_t p_offset_minutes) con
...
-	datetime[MINUTE_KEY] = time.minute;
-	datetime[SECOND_KEY] = time.second;
+	datetime[YEAR_KEY] = dt.year;
+	datetime[MONTH_KEY] = (uint8_t)dt.month;
+	datetime[DAY_KEY] = dt.day;
+	datetime[WEEKDAY_KEY] = (uint8_t)dt.weekday;
+	datetime[DST_KEY] = dt.dst;
+	datetime[HOUR_KEY] = dt.hour;
+	datetime[MINUTE_KEY] = dt.minute;
+	datetime[SECOND_KEY] = dt.second;
```
Comment (aaronfranke) [inline core/os/time.cpp:336]:
```suggestion
	datetime[HOUR_KEY] = dt.hour;
	datetime[MINUTE_KEY] = dt.minute;
	datetime[SECOND_KEY] = dt.second;
	datetime[DST_KEY] = dt.dst;
```

File: platform/windows/os_windows.cpp [hunk 4]
```diff
@@ -305,28 +305,16 @@ OS::Date OS_Windows::get_date(bool p_utc) const {
...
-	return time;
+	DateTime dt;
+	dt.day = systemtime.wDay;
+	dt.month = Month(systemtime.wMonth);
+	dt.weekday = Weekday(systemtime.wDayOfWeek);
+	dt.year = systemtime.wYear;
+	dt.dst = daylight;
+	dt.hour = systemtime.wHour;
+	dt.minute = systemtime.wMinute;
+	dt.second = systemtime.wSecond;
```
Comment (aaronfranke) [inline platform/windows/os_windows.cpp:316]:
```suggestion
	dt.hour = systemtime.wHour;
	dt.minute = systemtime.wMinute;
	dt.second = systemtime.wSecond;
	dt.dst = daylight;
```

File: platform/uwp/os_uwp.cpp [hunk 5]
```diff
@@ -444,36 +444,31 @@ String OS_UWP::get_name() const {
...
-	return time;
+	DateTime dt;
+	dt.day = systemtime.wDay;
+	dt.month = Month(systemtime.wMonth);
+	dt.weekday = Weekday(systemtime.wDayOfWeek);
+	dt.year = systemtime.wYear;
+	dt.dst = daylight;
+	dt.hour = systemtime.wHour;
+	dt.minute = systemtime.wMinute;
+	dt.second = systemtime.wSecond;
```
Comment (aaronfranke) [inline platform/uwp/os_uwp.cpp:470]:
```suggestion
	dt.hour = systemtime.wHour;
	dt.minute = systemtime.wMinute;
	dt.second = systemtime.wSecond;
	dt.dst = daylight;
```

File: platform/macos/export/export_plugin.cpp [hunk 6]
```diff
@@ -1686,16 +1685,15 @@ void EditorExportPlatformMacOS::_zip_folder_recursive(zipFile &p_zip, const Stri
...
-			zipfi.tmz_date.tm_min = time.minute;
-			zipfi.tmz_date.tm_mon = date.month - 1; // Note: "tm" month range - 0..11, Godot month range - 1..12, https://www.cplusplus.com/reference/ctime/tm/
-			zipfi.tmz_date.tm_sec = time.second;
-			zipfi.tmz_date.tm_year = date.year;
+			zipfi.tmz_date.tm_hour = dt.hour;
+			zipfi.tmz_date.tm_mday = dt.day;
+			zipfi.tmz_date.tm_min = dt.minute;
+			zipfi.tmz_date.tm_mon = dt.month - 1; // Note: "tm" month range - 0..11, Godot month range - 1..12, https://www.cplusplus.com/reference/ctime/tm/
+			zipfi.tmz_date.tm_sec = dt.second;
+			zipfi.tmz_date.tm_year = dt.year;
```
Comment (aaronfranke) [inline platform/macos/export/export_plugin.cpp:1696]:
```suggestion
			zipfi.tmz_date.tm_year = dt.year;
			zipfi.tmz_date.tm_mon = dt.month - 1; // Note: "tm" month range - 0..11, Godot month range - 1..12, https://www.cplusplus.com/reference/ctime/tm/
			zipfi.tmz_date.tm_mday = dt.day;
			zipfi.tmz_date.tm_hour = dt.hour;
			zipfi.tmz_date.tm_min = dt.minute;
			zipfi.tmz_date.tm_sec = dt.second;
```

File: platform/macos/export/export_plugin.cpp [hunk 7]
```diff
@@ -1641,16 +1641,15 @@ void EditorExportPlatformMacOS::_zip_folder_recursive(zipFile &p_zip, const Stri
...
```
Comment (aaronfranke) [inline platform/macos/export/export_plugin.cpp:1652]:
```suggestion
			zipfi.tmz_date.tm_year = dt.year;
			zipfi.tmz_date.tm_mon = dt.month - 1; // Note: "tm" month range - 0..11, Godot month range - 1..12, https://www.cplusplus.com/reference/ctime/tm/
			zipfi.tmz_date.tm_mday = dt.day;
			zipfi.tmz_date.tm_hour = dt.hour;
			zipfi.tmz_date.tm_min = dt.minute;
			zipfi.tmz_date.tm_sec = dt.second;
```

File: platform/android/export/export_plugin.cpp [hunk 8]
```diff
@@ -569,16 +569,15 @@ bool EditorExportPlatformAndroid::_should_compress_asset(const String &p_path, c
...
-	zipfi.tmz_date.tm_min = time.minute;
-	zipfi.tmz_date.tm_mon = date.month - 1; // tm_mon is zero indexed
-	zipfi.tmz_date.tm_sec = time.second;
-	zipfi.tmz_date.tm_year = date.year;
+	zipfi.tmz_date.tm_hour = dt.hour;
+	zipfi.tmz_date.tm_mday = dt.day;
+	zipfi.tmz_date.tm_min = dt.minute;
+	zipfi.tmz_date.tm_mon = dt.month - 1; // tm_mon is zero indexed
+	zipfi.tmz_date.tm_sec = dt.second;
+	zipfi.tmz_date.tm_year = dt.year;
```
Comment (aaronfranke) [inline platform/android/export/export_plugin.cpp:580]:
```suggestion
	zipfi.tmz_date.tm_year = dt.year;
	zipfi.tmz_date.tm_mon = dt.month - 1; // tm_mon is zero indexed
	zipfi.tmz_date.tm_mday = dt.day;
	zipfi.tmz_date.tm_hour = dt.hour;
	zipfi.tmz_date.tm_min = dt.minute;
	zipfi.tmz_date.tm_sec = dt.second;
```

**Review by aaronfranke — COMMENTED** [review]
One more minor thing I noticed, we should make the order consistent.

**Review by aaronfranke — APPROVED** [review]

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**